### DISCUSSION TITLE
Welcome to cadical Discussions!

### DISCUSSION THREAD

**Discussion by arminbiere** [arminbiere/cadical#58] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by MxaWnag** [comment]
I am a new comer for solvers, and now I want to connect the latest CaDiCal to Painless V2  for experiments, and I want to export and import learning clause in the cdcl loop, what api should I focus on？

**Reply by m-fleury** [reply]
It slightly depends on what you want to achieve, but I am not sure that we offer exactly the interface we need:

* IPASIR interface + learner
   - if you want to run CaDiCaL and interrupt it (let's say every 10s), then you can use the ipasir interface, set the option to interrupt the solver, and use the learner interface that iterates over all clauses.
   - But: this approach means that the solver will wait. And the solver will undo the inprocessing if you add a clause with a variable that was eliminated.

* IPASIR-UP interface
   - or you could go with the IPASIR-UP interface. Here you can ignore all the assignment notifications, you only care about adding and removing clauses.
   - But: the ipasir-up interface assumes that whenever you add a clause during search it will only be over observed variables (aka frozen), which is probably not what you want (maybe?).

**Reply by m-fleury** [reply]
and I forgot the proof tracer to get the learned clauses, but you only get the clauses, no other information (like glue or used flag).

But you probably need some extensions on top of the existing API.

**Reply by MxaWnag** [reply]
Thank you for your reply.  actually i have read the Api of IPASIR, and i have used the Learner interface for obtaining all clauses. 
 My purpose is  to export and import learned clauses during cdcl loop, and i have got the learned clause for each conflict,but for importing,i used Api called "new_learned_redundant_clause(int glue)" and sent it a determined glue value.In my understanding，it will automatically add the clause but it doesn`t seem to work

**Reply by MxaWnag** [reply]
And i haven`t tried IPAIR-UP yet, maybe it can work

**Reply by m-fleury** [reply]
`new_learned_redundant_clause` is only an internal API. It requires that the content is put into `clause`, but it requires that the clause is correctly sorted for watching. For sure not something you want to use. And it is done after compacting (aka renaming the variables).

**Reply by MxaWnag** [reply]
Thank you for reply. So maybe the IPASIR-UP interface is actually what i want, but it seems that the interface can not determind a glue for a clause i want to add

**Reply by m-fleury** [reply]
[I cannot find that code in the latest release](https://github.com/arminbiere/cadical/blob/master/src/internal.cpp#L248). To use the latest CaDiCaL version as unchanged as possible, then yes, IPASIR-UP is the way to go.

**Reply by MxaWnag** [reply]
yeah it can't be found in latest cadical  because  it's  in my  
individual version.The callback are  functions I added later for  experiments. If the  interface can  be used then  I am going to try it

**Reply by m-fleury** [reply]
Try to see what you need API-wise and we will discuss internally if we want to support this. For example, we could have a way to ignore the clause you are adding if it contains eliminated variables -- currently this is a violation of the contract. This is not useful for the current usage of the user-propagator (@kfazekas can you confirm?) but this can be useful for parallel SAT solving.

**Reply by MxaWnag** [reply]
thank  you  for your suggestion on using ipasir-up. I  have  connected  cadical to painless successfully by implementing  Learner  and  External Propagator ,but there is still room for improvement.It would  be much better  if  the  interface can support handling  with the information  of shared  clauses like  lbd value

**Reply by m-fleury** [reply]
On the one hand, I understand your motivation to have it. On the other hand, I am not aware of any other application that requires this knowledge and it is extremely constraining as it means that have to support this and cannot change the meaning anymore (like let say replacing glue by lbd which is one off).


How did you solve the issue of variable elimination?

**Reply by MxaWnag** [reply]
I realized that I seem to have made some mistakes, ExternalPropagator doesn't seem to be able to be used to import shared learning clauses directly, you have to import propagated_elit first, is there a way to import learning clauses directly.

And i didn`t solve the issue of variable elimination.

**Reply by m-fleury** [reply]
The comment above `cb_has_external_clause` indicates that it should work. What is your issue exactly?

**Reply by MxaWnag** [reply]
I have tried cb_has_external_clause  And  cb_add_external_clause_lit  but it didn`t work well, and i tried to use cb_add_reason_clause_lit .Maybe i am misunderstanding the concept, Is it enough to use cb_has_external_clause And cb_add_external_clause_lit if i want to import learning clause from other solvers?
